File: packages/node/motia-example/steps/todo/update-todo.step.ts
import { ApiRouteConfig, Handlers } from 'motia'
import { z } from 'zod'
import { todoSchema } from './create-todo.step'

export const config: ApiRouteConfig = {
  type: 'api',
  name: 'UpdateTodo',
  description: 'Update an existing todo item',
  flows: ['todo-app'],

  method: 'PUT',
  path: '/todo/:todoId',
  bodySchema: z.object({
    description: z.string().optional(),
    dueDate: z.string().optional(),
    checked: z.boolean().optional(),
  }),
  responseSchema: {
    200: todoSchema,
    404: z.object({ error: z.string() }),
  },
  emits: [],
  virtualSubscribes: ['todo-created'],
}

export const handler: Handlers['UpdateTodo'] = async (req, { logger, streams }) => {
  const { todoId } = req.pathParams
  logger.info('Updating todo', { todoId, body: req.body })

  const existingTodo = await streams.todo.get('inbox', todoId)

  if (!existingTodo) {
    logger.warn('Todo not found', { todoId })

    return {
      status: 404,
      body: { error: `Todo with id ${todoId} not found` },
    }
  }

  const updatedTodo = { ...existingTodo }

  if (req.body.checked !== undefined) {
    updatedTodo.completedAt = req.body.checked ? new Date().toISOString() : undefined
  }

  if (req.body.description !== undefined) {
    updatedTodo.description = req.body.description
  }

  if (req.body.dueDate !== undefined) {
    updatedTodo.dueDate = req.body.dueDate
  }

  const result = await streams.todo.set('inbox', todoId, updatedTodo)

  logger.info('Todo updated successfully', { todoId })

  return { status: 200, body: result }
}

File: packages/python/motia-example/steps/todo/update_todo.step.py
"""Update Todo Step."""

from typing import Any

from motia import ApiResponse, ApiTrigger, FlowContext, StepConfig, Stream, TriggerInput, step_wrapper

todo_stream: Stream[dict[str, Any]] = Stream("todo")


config = StepConfig(
    name="UpdateTodo",
    description="Update an existing todo item",
    flows=["todo-app"],
    triggers=[
        ApiTrigger(path="/todo", method="PUT"),
    ],
    emits=[],
)


async def handler(input: TriggerInput[dict[str, Any]], ctx: FlowContext[Any]) -> ApiResponse[Any]:
    """Handle update todo request."""
    body = input.data or {}
    todo_id = body.get("todo_id")

    ctx.logger.info("Updating todo", body)

    if not todo_id:
        ctx.logger.error("todo_id is required")
        return ApiResponse(status=400, body={"error": "todo_id is required"})

    existing_todo = await todo_stream.get("inbox", todo_id)

    if not existing_todo:
        ctx.logger.error("Todo not found")
        return ApiResponse(status=404, body={"error": "Todo not found"})

    # Merge existing todo with updates
    updated_todo = {**existing_todo, **body}
    todo = await todo_stream.set("inbox", todo_id, updated_todo)

    ctx.logger.info("Todo updated successfully", {"todo_id": todo_id})

    return ApiResponse(status=200, body=todo)


step_wrapper(config, __file__, handler)
